// Copyright 2021, Yang Luo"
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// @Author
// Yang Luo, PHD
// Shenyang Institute of Automation, Chinese Academy of Sciences.
// email: luoyang@sia.cn

syntax = "proto3";

package rocos;

option java_outer_classname = "RobotInfoProto";

import "header.proto"; // Standard header of gRPC
import "geometry.proto";
// import "google/protobuf/wrappers.proto";
// import "google/protobuf/timestamp.proto";

message JointInfo {
  string name = 1; // 关节名称
  double cnt_per_unit = 2; // 每个用户单位对应多少脉冲数
  double torque_per_unit = 3; // 每个用户单位对应多少Torque值
  double ratio = 4; // 关节减速比
  int32  pos_zero_offset = 5; //关节零位偏移
  string user_unit_name = 6; // 用户单位的名称，可能是rad, m, mm等
}

enum JointType
{
  UNKNOWN = 0;
  REVOLUTE = 1;
  CONTINUOUS = 2;
  PRISMATIC = 3;
  FLOATING = 4;
  PLANAR = 5;
  FIXED = 6;
}

message Link {
  string name = 1; // Link Name
  int32 order = 2; // Link Order
  JointType type = 3; // Link Type(fixed, continuous, revolute, prismatic)
  Vec3 translate = 4; // joint origin xyz
  Vec3 rotate = 5; // joint origin rpy
  Vec3 axis = 6; // joint axis
  Vec3 translateLink = 7; // link origin xyz
  Vec3 rotateLink = 8; // link origin rpy
  string mesh = 9; // link mesh path
}

message RobotModel {
  string name = 1; // robot name
  repeated Link links = 2; // all links
}

message LinkMeshPath { // mesh file path in rocos-app
  string path = 1; // mesh file path
}

message LinkMeshFile{
  string name = 1; // mesh file name
  bytes  content = 2; // mesh file content
}

// The Infomation of the robot
message RobotInfo {
  repeated JointInfo joint_infos = 1; //包含所有关节信息
}

message RobotInfoRequest {
  RequestHeader header = 1; // Common request header.
}

message RobotInfoResponse {
  ResponseHeader header = 1; //Common request header.
  RobotInfo robot_info = 2; // 机器人信息
}